File: src/main/java/ui_adapter/model/Selector.java
package ui_adapter.model;

public class Selector {
    public enum Type {
        CSS,
        XPATH,
        TEXT,
        ROLE,
        LABEL,
        PLACEHOLDER,
        TEST_ID
    }

    private final Type type;
    private final String value;

    // Optional semantic metadata to help the executor build robust locators
    private final String role;        // e.g. button, link, textbox, checkbox
    private final String nearText;    // nearby text to disambiguate
    private final String description; // free-form description from the LLM

    public Selector(Type type, String value) {
        this(type, value, null, null, null);
    }

    public Selector(Type type, String value, String role, String nearText, String description) {
        this.type = type;
        this.value = value;
        this.role = role;
        this.nearText = nearText;
        this.description = description;
    }

    public Type getType() {
        return type;
    }

    public String getValue() {
        return value;
    }

    public String getRole() {
        return role;
    }

    public String getNearText() {
        return nearText;
    }

    public String getDescription() {
        return description;
    }

    @Override
    public String toString() {
        return String.format("%s=%s (role=%s, nearText=%s, description=%s)",
                type, value, role, nearText, description);
    }
}

File: python-code-base/playwright_ui_adapter/model/selector.py
"""Selector model for UI element location strategies."""

from dataclasses import dataclass
from enum import Enum
from typing import Optional


class SelectorType(str, Enum):
    """Selector strategy types."""
    
    CSS = "CSS"
    XPATH = "XPATH"
    TEXT = "TEXT"
    ROLE = "ROLE"
    LABEL = "LABEL"
    PLACEHOLDER = "PLACEHOLDER"
    TEST_ID = "TEST_ID"
    
    def __str__(self) -> str:
        return self.value


@dataclass
class Selector:
    """
    Represents a selector for locating UI elements.
    
    Supports multiple selector strategies with optional metadata
    for disambiguation and semantic understanding.
    
    Attributes:
        type: The selector strategy type
        value: The selector value (e.g., CSS selector, text content)
        role: Optional ARIA role for ROLE strategy
        near_text: Optional nearby text for disambiguation
        description: Optional LLM-provided description
    """
    
    type: SelectorType
    value: str
    role: Optional[str] = None
    near_text: Optional[str] = None
    description: Optional[str] = None
    
    def __str__(self) -> str:
        parts = [f"{self.type}={self.value}"]
        if self.role:
            parts.append(f"role={self.role}")
        if self.near_text:
            parts.append(f"near={self.near_text}")
        if self.description:
            parts.append(f"desc={self.description[:30]}...")
        return f"Selector({', '.join(parts)})"
    
    def __repr__(self) -> str:
        return self.__str__()
